### DISCUSSION TITLE
NFT: on opensea, I can only search my minted NFTs by smart contract name, but the CID of the NFTs (neither text data nor image CID can be searched)

### DISCUSSION THREAD

**Discussion by Vincent71399** [Cyfrin/foundry-full-course-cu#2484] [Q&A]
I searched on https://testnets.opensea.io/ since I deployed my contract to Sepolia
is it the way it should work?

another question is, how to prevent the same image being minted by two different smart contracts?

**Comment by EngrPips** [comment]
I think an opensea search works by contract address, and if you want to search for a `CID`, you will need to send a request to an IPFS network. I think you are responsible for setting the image of your NFT, so by default, two different contracts are not meant to mint the same image except if you have set the same image in the two contracts.

**Reply by Vincent71399** [reply]
what if someone call getTokenURI of my contract to retrieve my image CID of IPFS? then write another contract to mint them?

**Reply by EngrPips** [reply]
I think I get the gist now. The way Patrick designed the NFT contract in the course isn't the traditional way of designing an NFT contract, normally the minter doesn't always have access to what their minted NFT will look like and like that, it's hard to intentionally mint the same image, but the course is for learning purpose and we can do whatever we like. If you have minted an NFT before did you provide the image of the NFT when you minted?

**Reply by Vincent71399** [reply]
Let's consider a scenario,  there is an NFT minting contract A, user1 mint a few NFTs (NFT1, NFT2, NFT3) on it, so user1 is the owner of these 3 NFTs and definitely has access to those image/meta URIs. user1 then create another NFT smart contract B, where allows other users to mint these 3 NFTs, but even if they minted in the contract B, these NFTs are still belong to user1 in contract A, please correct me if I am wrong

**Reply by EngrPips** [reply]
Do you realize that an NFT smart contract keeps a record of an address to the tokenId it owns? Anybody can have an NFT picture, which has no meaning. What matters when you want to know who owns which NFT is to check the mapping on the NFT smart contract and see which address is mapped to a particular tokenId. That is how you know the owner of an NFT.

**Reply by Vincent71399** [reply]
that's my question, it is possible that contract A map an URL to address1, while contract B map the same URL to address2, then who should be the owner of the image behind the URL?

**Reply by EngrPips** [reply]
Well, that is why when you want to mint an NFT, you are not always the one that provides the Image of your NFT or have you minted an NFT before where you are the one that provides the url of your NFT? and you are taking a mere tutorial code too seriously. The codebase is meant to teach you about ERC721, and thats all about it.

**Reply by Vincent71399** [reply]
In real world, what would be the solution?

**Reply by EngrPips** [reply]
In the real world, this problem doesn't exist because the user doesn't pass in the `url` of their token when minting, but the protocol determines what the `url` will be, and there is no two tokenId with identical `url.` Does this make sense to you?

**Reply by Vincent71399** [reply]
> In the real world, this problem doesn't exist because the user doesn't pass in the `url` of their token when minting, but the protocol determines what the `url` will be, and there is no two tokenId with identical `url.` Does this make sense to you?

Is it possible if a hacker NFT contract copys urls from a normal NFT contract?

**Reply by EngrPips** [reply]
At least that is a case where one protocol is original and the other is fake, that's a whole different scenario

**Reply by Vincent71399** [reply]
I think you misunderstood my question, in two different contracts, it is possible for the same url to be minted. So the same url can belongs to different owners in each contract of these two.

**Reply by EngrPips** [reply]
Why will two contract point to the same url for a tokenId?

**Reply by EngrPips** [reply]
To make it easy for both of us bring me two NFT protocol that have the same tokenId point to the same `URL to make it easier for you I will take it if different tokenId on the two protocol points to the same url
